examples/LC.hs

```diff
@@ -207,64 +209,146 @@ nf (App e1 e2) =
+-- In the case of beta-reduction, the `unBindWith` operation
 -- applies its argument to the environment and subterm in the
--- closure. In other words, this function calls `evalEnv` 
+-- closure. In other words, this function calls `evalEnv`
 -- recursively with the saved environment and body of the lambda term.
--- Because `evalEnv` takes the body of the lambda term directly, 
--- without substitution, it doesn't do any repeat work. 
+-- Because `evalEnv` takes the body of the lambda term directly,
+-- without substitution, it doesn't do any repeat work.
 
 evalEnv :: Env Exp m n -> Exp m -> Exp n
 evalEnv r (Var x) = applyEnv r x
 evalEnv r (Lam b) = applyE r (Lam b)
 evalEnv r (App e1 e2) =
-    let v = evalEnv r e2 in
-    case evalEnv r e1 of
-        Lam b -> 
-            unbindWith (\r' e' -> evalEnv (v .: r') e') b
+  let v = evalEnv r e2
+   in case evalEnv r e1 of
+        Lam b ->
+          unbindWith (\r' e' -> evalEnv (v .: r') e') b
         t -> App t v
 
-
 -- >>> evalEnv zeroE t1     -- start with "empty environment"
 -- λ. λ. 1 (λ. 0 0)
 
-
 -- For full reduction, we need to normalize under the binder too.
--- In this case, the `applyUnder` function takes care of the 
+-- In this case, the `applyUnder` function takes care of the
 -- necessary environment manipulation. It applies its argument (`nfEnv`)
--- to the modifed 
+-- to the modifed
 
 -- >>> :t applyUnder nfEnv
 -- applyUnder nfEnv :: Env Exp n1 n2 -> Bind Exp Exp n1 -> Bind Exp Exp n2
 --
 -- In the beta-reduction case, we could use `unbindWith` as above
 -- but the `instantiateWith` function already captures exactly
--- this pattern. 
+-- this pattern.
 nfEnv :: Env Exp m n -> Exp m -> Exp n
 nfEnv r (Var x) = applyEnv r x
 nfEnv r (Lam b) = Lam (applyUnder nfEnv r b)
 nfEnv r (App e1 e2) =
-    let n = nfEnv r e1 in
-    case nfEnv r e1 of
+  let n = nfEnv r e1
+   in case nfEnv r e1 of
         Lam b -> instantiateWith nfEnv b n
         t -> App t (nfEnv r e2)
 
 ----------------------------------------------------------------
+
+t2 = Lam (bind (App (Lam (bind (Lam (bind (Var f0))))) (Var f0)))
+
+-- >>> t2
+-- (λ. ((λ. (λ. 0)) 0))
+
+t3 = Lam (bind (App (App (Lam (bind (Var f0))) (Lam (bind (Var f0)))) (App (Lam (bind (Var f0))) (Lam (bind (Var f0))))))
+
+-- >>> t3
+-- (λ. (((λ. 0) (λ. 0)) ((λ. 0) (λ. 0))))
+
+t4 = Lam (bind (App (Var f0) (Lam (bind (Var f1)))))
+
+-- >>> t4
+-- (λ. (0 (λ. 1)))
+
+t5 = Lam (bind (App (Lam (bind (Var f0))) (App (Var f0) (Lam (bind (Var f1))))))
+
+-- >>> t5
+-- (λ. ((λ. 0) (0 (λ. 1))))
+
+betaEqual :: Exp m -> Exp m -> Bool
+betaEqual a b = nf a == nf b
+
+-- >>> betaEqual t0 (nf t0)
+-- True
+-- >>> betaEqual t1 (nf t1)
+-- True
+-- >>> betaEqual t2 t3
+-- True
+-- >>> betaEqual t4 t5
+-- True
+
+reduceApply :: Exp m -> Exp m -> Exp m
+reduceApply a b =
+  case a of
+    Lam a -> instantiate a b
+    _ -> App a b
+
+shortCircuitEq' :: Exp m -> Exp m -> (Bool, Exp m, Exp m)
```

Comment:
This looks similar to what I was thinking about. Checkout my version in the PTS example: https://github.com/sweirich/autoenv/blob/01323784b89d09cb1419f1eca884668a101e71de/examples/PTS.hs#L347

Your "reduceApply" is one step of the "whnf" function in my version, so your version interleaves reduction even more than mine does. I wonder which version is faster?